File: DSAlgorithms/Leetcode/c++/binary_search/2300-spells_and_pots.cpp
class Solution {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
       sort(potions.begin(), potions.end());
       std::vector<int > result;
       for(int i = 0; i < spells.size(); i++)
       {
            int left = 0;
            int right = potions.size() - 1;
            int index = potions.size();
            while(left <= right) 
            {
                int mid = left + (right - left) / 2;
                if((double)spells[i] * potions[mid] >= success)
                {
                    right = mid - 1;
                    index = mid;
                }
                else
                    left = mid + 1;
            }
            result.push_back(potions.size() - index);
       }
       return result;
    }
};

File: dsa_algo/leetcode/ts/src/binary_search/2300-spells_and_pots.ts
const successfulPairs = (spells: number[], potions: number[], success: number): number[] =>{
    potions.sort((a, b) => a - b);
    const result : number[] = [];
    for(let i = 0; i < spells.length; i++)
    {
        let left = 0;
        let right = potions.length - 1;
        let index = potions.length; 
        while(left <= right)
        {
            let mid = Math.floor(left + (right - left) /2);
            if(spells[i] * potions[mid] >= success)
            {
                right = mid - 1;
                index = mid;
            }
            else 
                left = mid + 1;
        }
        result.push(potions.length - index);
    }
    return result;
};